spot_driver/spot_driver/calibrated_reregistered_hand_camera_depth_publisher.py

```diff
@@ -0,0 +1,233 @@
+
+
+class CalibratedReRegisteredHandCameraDepthPublisher:
+    def __init__(
+        self,
+        calibration_path: Optional[str] = None,
+        tag: str = "default",
+        robot_name: str = "spot",
+        topic_name: str = "depth_registered/hand_custom_cal/image",
+        undistort: bool = False,
+    ):
+        self.node = ros_process.node()
+
+        if calibration_path is None:
+            self.node.get_logger().warning("No calibration path found, using default calibration")
+            self.calibration = {
+                "camera_matrix_depth": np.zeros(9).tolist(),
+                "dist_coeffs_depth": np.zeros(5).tolist(),
+                "camera_matrix_rgb": np.zeros(9).tolist(),
+                "dist_coeffs_rgb": np.zeros(5).tolist(),
+                "depth_t_rgb_R": np.zeros(9).tolist(),
+                "depth_t_rgb_T": np.zeros(3).tolist(),
+                "rgb_image_dim": np.zeros(2).tolist(),
+            }
+        else:
+            self.calibration = extract_calibration_parameters(calibration_path=calibration_path, tag=tag)
+
+        self.calibration["robot_name"] = robot_name
+        self.calibration["topic_name"] = topic_name
+        self.calibration["undistort"] = undistort
+
+        for param_name, param_val in self.calibration.items():
+            self.node.get_logger().info(f"Setting the parameter {param_name} to a default value of {param_val}")
+            self.node.declare_parameter(param_name, param_val)
+            self.calibration[param_name] = self.node.get_parameter(param_name).value
+            if param_name in ["camera_matrix_depth", "camera_matrix_rgb", "depth_t_rgb_R"]:
+                self.calibration[param_name] = np.array(self.calibration[param_name]).reshape((3, 3))
+            elif "dim" not in param_name and "name" not in param_name and "undistort" not in param_name:
+                self.calibration[param_name] = np.array(self.calibration[param_name])
+
+        self.cv_bridge = CvBridge()
+
+        raw_depth_topic = f"/{self.calibration['robot_name']}/depth/hand/image"
+        self.node.get_logger().info(f"Creating subscriber to raw depth at {raw_depth_topic}")
+        self.raw_depth_img_sub = self.node.create_subscription(
+            Image, raw_depth_topic, self.republish_registered_depth_callback, 10
+        )
+
+        reregistered_depth_topic = f"/{self.calibration['robot_name']}/{topic_name}"
+        self.node.get_logger().info(f"Creating reregistered depth publisher to {reregistered_depth_topic}")
+        self.reregistered_depth_img_pub = self.node.create_publisher(Image, reregistered_depth_topic, 10)
+
+    def republish_registered_depth_callback(self, msg: Image) -> None:
+        if self.calibration["rgb_image_dim"][0] != 0:
+            raw_depth_image = self.cv_bridge.imgmsg_to_cv2(msg, desired_encoding="passthrough")
+            pointcloud_pcd = self.depth_img_to_pointcloud(raw_depth_image, undistort=self.calibration["undistort"])
```

Comment:
Should we also publish this point cloud?

Reply:
The point cloud can easily be composed independent of this republisher in the same manner that they are for the default hand. We could ROS launch the depth_image_proc xyzrgb node to do this without having to implement the projection ourselves.

I'm also down to just publish the point cloud here, but I think using depth_image_proc for this as well makes sense.

I didn't include this off of the bat for a similar reasoning as the launch file(https://github.com/bdaiinstitute/spot_ros2/pull/455#discussion_r1722328087). If we want to include the republisher in the launch, we can also put the depth_image_proc node in as well

Reply:
Agree with using depth_image_proc down the line, would probably also be faster as it's using C++

Reply:
I guess I was just thinking if we already generate the point cloud here then why start another process to generate the point cloud

Reply:
@glvov-bdai Is there anything else that would need to be modified with `point_cloud_pcd` in order to publish it? Besides converting it to ROS and giving it a stamp, etc. 
If there's no additional processing we would need to do, I agree that it makes sense to just publish it here. (I was assuming above that there might be other steps but that could be wrong)

Reply:
@khughes-bdai Before the point cloud can be used downstream, it must be colorized as well. If we want to colorize the pointcloud, we need to also subscribe to the RGB image stream (we only subscribe to raw depth here)

It may make sense to do so as we could also undistort the rgb image here based off of the loaded custom calibration intrinsic. 

However, this would require adding a time synchronizer for the RGB and depth images. I agree with @amessing-bdai that it makes sense to reduce the number of processes started but it may also increase the complexity of the publisher.

I also wonder if maybe spawning another process to generate the clouds could actually boost performance, as clouds are colorized and populated for publication parallel to/independent of creating the registered depth image (and without adding rgb rectification to the republishing loop).

I am impartial to either adding it as an depth_image_proc node, publishing it directly here from open3d, or maybe not even publishing it as we already publish the depth image. Let me know what we like best ;-p

Reply:
In that case, I think start another depth image proc makes sense to not add to the complexity of this node. That can also be in the future PR that adds to the launch file.